File: Game-1-modular/data/models/skills.py
"""Skill-related data models"""

from dataclasses import dataclass, field
from typing import List, Dict, Optional


@dataclass
class SkillEffect:
    """Represents a skill's effect"""
    effect_type: str  # empower, quicken, fortify, etc.
    category: str  # mining, combat, smithing, etc.
    magnitude: str  # minor, moderate, major, extreme
    target: str  # self, enemy, area, resource_node
    duration: str  # instant, brief, moderate, long, extended
    additional_effects: List[Dict] = None

    def __post_init__(self):
        if self.additional_effects is None:
            self.additional_effects = []


@dataclass
class SkillCost:
    """Represents skill costs"""
    mana: str  # low, moderate, high, extreme
    cooldown: str  # short, moderate, long, extreme


@dataclass
class SkillEvolution:
    """Represents skill evolution data"""
    can_evolve: bool
    next_skill_id: Optional[str]
    requirement: str


@dataclass
class SkillRequirements:
    """Represents skill requirements"""
    character_level: int
    stats: Dict[str, int]
    titles: List[str]


@dataclass
class SkillDefinition:
    """Complete skill definition from JSON"""
    skill_id: str
    name: str
    tier: int
    rarity: str
    categories: List[str]
    description: str
    narrative: str
    tags: List[str]
    effect: SkillEffect
    cost: SkillCost
    evolution: SkillEvolution
    requirements: SkillRequirements
    icon_path: Optional[str] = None


@dataclass
class PlayerSkill:
    """Player's learned skill instance"""
    skill_id: str
    level: int = 1
    experience: int = 0
    current_cooldown: float = 0.0  # Cooldown remaining in seconds
    is_equipped: bool = False  # Whether this skill is equipped to hotbar
    hotbar_slot: Optional[int] = None  # Which hotbar slot (1-5) or None

    def get_definition(self) -> Optional['SkillDefinition']:
        """Get the full definition from SkillDatabase"""
        from data.databases.skill_db import SkillDatabase
        db = SkillDatabase.get_instance()
        return db.skills.get(self.skill_id, None)

    def get_exp_for_next_level(self) -> int:
        """Get EXP required to reach next level (exponential doubling)"""
        if self.level >= 10:
            return 0  # Max level
        # Level 1→2 = 1000, Level 2→3 = 2000, Level 3→4 = 4000, etc.
        return 1000 * (2 ** (self.level - 1))

    def add_exp(self, amount: int) -> tuple:
        """
        Add skill EXP and check for level up.
        Returns (leveled_up, new_level)
        """
        if self.level >= 10:
            return False, self.level  # Max level

        self.experience += amount
        leveled_up = False
        old_level = self.level

        # Check for level ups (can level multiple times if enough EXP)
        while self.level < 10:
            exp_needed = self.get_exp_for_next_level()
            if self.experience >= exp_needed:
                self.experience -= exp_needed
                self.level += 1
                leveled_up = True
                print(f"   🌟 Skill Level Up! {self.skill_id} → Level {self.level}")
            else:
                break

        return leveled_up, self.level if leveled_up else old_level

    def get_level_scaling_bonus(self) -> float:
        """Get effectiveness bonus from skill level (+10% per level)"""
        return 0.1 * (self.level - 1)  # Level 1 = +0%, Level 10 = +90%

    def can_use(self) -> bool:
        """Check if skill is off cooldown"""
        return self.current_cooldown <= 0

    def update_cooldown(self, dt: float):
        """Update cooldown timer"""
        if self.current_cooldown > 0:
            self.current_cooldown = max(0, self.current_cooldown - dt)

    def start_cooldown(self, cooldown_seconds: float):
        """Start cooldown timer"""
        self.current_cooldown = cooldown_seconds

File: Unity/Assets/Scripts/Game1.Data/Models/Skills.cs
// Game1.Data.Models.Skills
// Migrated from: data/models/skills.py (135 lines)
// Phase: 1 - Foundation
// Contains: SkillEffect, SkillCost, SkillEvolution, SkillRequirements, SkillDefinition, PlayerSkill

using System;
using System.Collections.Generic;
using Newtonsoft.Json;

namespace Game1.Data.Models
{
    [Serializable]
    public class SkillEffect
    {
        [JsonProperty("effectType")]
        public string EffectType { get; set; }

        [JsonProperty("category")]
        public string Category { get; set; }

        [JsonProperty("magnitude")]
        public string Magnitude { get; set; }

        [JsonProperty("target")]
        public string Target { get; set; }

        [JsonProperty("duration")]
        public string Duration { get; set; }

        [JsonProperty("additionalEffects")]
        public List<Dictionary<string, object>> AdditionalEffects { get; set; } = new();
    }

    [Serializable]
    public class SkillCost
    {
        /// <summary>
        /// Mana cost - supports both string ("low", "moderate", "high", "extreme")
        /// and numeric values (20-150). Stored as object for Union compatibility.
        /// </summary>
        [JsonProperty("mana")]
        public object Mana { get; set; }

        /// <summary>
        /// Cooldown - supports both string ("short", "moderate", "long", "extreme")
        /// and numeric values (10-600). Stored as object for Union compatibility.
        /// </summary>
        [JsonProperty("cooldown")]
        public object Cooldown { get; set; }

        public float GetManaAsFloat()
        {
            if (Mana is double d) return (float)d;
            if (Mana is long l) return l;
            if (Mana is int i) return i;
            if (Mana is string s && float.TryParse(s, out float f)) return f;
            return 0f;
        }

        public float GetCooldownAsFloat()
        {
            if (Cooldown is double d) return (float)d;
            if (Cooldown is long l) return l;
            if (Cooldown is int i) return i;
            if (Cooldown is string s && float.TryParse(s, out float f)) return f;
            return 0f;
        }
    }

    [Serializable]
    public class SkillEvolution
    {
        [JsonProperty("canEvolve")]
        public bool CanEvolve { get; set; }

        [JsonProperty("nextSkillId")]
        public string NextSkillId { get; set; }

        [JsonProperty("requirement")]
        public string Requirement { get; set; }
    }

    [Serializable]
    public class SkillRequirements
    {
        [JsonProperty("characterLevel")]
        public int CharacterLevel { get; set; }

        [JsonProperty("stats")]
        public Dictionary<string, int> Stats { get; set; } = new();

        [JsonProperty("titles")]
        public List<string> Titles { get; set; } = new();
    }

    [Serializable]
    public class SkillDefinition
    {
        [JsonProperty("skillId")]
        public string SkillId { get; set; }

        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("tier")]
        public int Tier { get; set; }

        [JsonProperty("rarity")]
        public string Rarity { get; set; }

        [JsonProperty("categories")]
        public List<string> Categories { get; set; } = new();

        [JsonProperty("description")]
        public string Description { get; set; }

        [JsonProperty("narrative")]
        public string Narrative { get; set; }

        [JsonProperty("tags")]
        public List<string> Tags { get; set; } = new();

        [JsonProperty("effect")]
        public SkillEffect Effect { get; set; }

        [JsonProperty("cost")]
        public SkillCost Cost { get; set; }

        [JsonProperty("evolution")]
        public SkillEvolution Evolution { get; set; }

        [JsonProperty("requirements")]
        public SkillRequirements Requirements { get; set; }

        [JsonProperty("iconPath")]
        public string IconPath { get; set; }

        [JsonProperty("combatTags")]
        public List<string> CombatTags { get; set; } = new();

        [JsonProperty("combatParams")]
        public Dictionary<string, object> CombatParams { get; set; } = new();
    }

    /// <summary>
    /// Player's learned skill instance with level progression.
    /// Critical Constants:
    ///   Max skill level: 10
    ///   EXP formula: 1000 * 2^(level - 1)
    ///   Level scaling bonus: 10% per level above 1
    /// </summary>
    [Serializable]
    public class PlayerSkill
    {
        [JsonProperty("skillId")]
        public string SkillId { get; set; }

        [JsonProperty("level")]
        public int Level { get; set; } = 1;

        [JsonProperty("experience")]
        public int Experience { get; set; }

        [JsonProperty("currentCooldown")]
        public float CurrentCooldown { get; set; }

        [JsonProperty("isEquipped")]
        public bool IsEquipped { get; set; }

        [JsonProperty("hotbarSlot")]
        public int? HotbarSlot { get; set; }

        /// <summary>
        /// Get EXP required for next level. Formula: 1000 * 2^(level - 1).
        /// Level 1->2 = 1000, 2->3 = 2000, ..., 9->10 = 256000.
        /// Returns 0 at max level (10).
        /// </summary>
        public int GetExpForNextLevel()
        {
            if (Level >= 10) return 0;
            return 1000 * (1 << (Level - 1)); // 1 << n == 2^n
        }

        /// <summary>
        /// Add skill EXP and check for level up. Can level multiple times.
        /// Returns (leveled_up, new_level).
        /// </summary>
        public (bool LeveledUp, int NewLevel) AddExp(int amount)
        {
            if (Level >= 10) return (false, Level);

            Experience += amount;
            bool leveledUp = false;
            int oldLevel = Level;

            while (Level < 10)
            {
                int expNeeded = GetExpForNextLevel();
                if (Experience >= expNeeded)
                {
                    Experience -= expNeeded;
                    Level++;
                    leveledUp = true;
                }
                else
                {
                    break;
                }
            }

            return (leveledUp, leveledUp ? Level : oldLevel);
        }

        /// <summary>
        /// Get effectiveness bonus from skill level (+10% per level).
        /// Level 1 = +0%, Level 10 = +90%.
        /// </summary>
        public float GetLevelScalingBonus() => 0.1f * (Level - 1);

        public bool CanUse() => CurrentCooldown <= 0;

        public void UpdateCooldown(float dt)
        {
            if (CurrentCooldown > 0)
                CurrentCooldown = Math.Max(0, CurrentCooldown - dt);
        }

        public void StartCooldown(float cooldownSeconds)
        {
            CurrentCooldown = cooldownSeconds;
        }
    }
}
